facebook/docusaurus
# feat(docs,theme-classic): docs breadcrumbs

## Motivation

There is a desire to have breadcrumbs visible in docs.

This PR introduces breadcrumbs in a few ways:

- `useDocsBreadcrumbs` hook - parses the current sidebar items for the active "path" of items, returning them in an array of objects containing `label` and optional `href`.
  - Uses roughly the same approach that @Josh-Cena pitched in the original [issue](https://github.com/facebook/docusaurus/issues/6096) to extract the breadcrumbs.
- `breadcrumbs` theme configuration property - accepts a boolean value, or `"nested"` to indicate that breadcrumbs should only be visible when not in a top-level page.
  - Defaults to `"nested"` (open to pushback on this, I just found it to be redundant to see a single breadcrumb item)
- `DocBreadcrumbs` component - utilizes the new hook to retrieve breadcrumb items, rendering them in a navigation element.
  - I added this component to `DocItem` and `DocCategoryGeneratedIndexPage`.
  - Uses the Infima [Breadcrumbs](https://infima.dev/docs/components/breadcrumbs/) component styles, with some slight modification to type size and default background color.
  - Takes the theme configuration value into consideration. If `false` never renders, if `"nested"` only renders if you are not on a top level page.
  - Renders items without `href`s as `<span>`s.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Here's how you can test this new feature.

- By default the theme config value is set to `"nested"`, so you should only see breadcrumbs on non top-level doc pages.
  - Setting this value to `true` should display breadcrumbs on all doc pages.
  - Setting this value to `false` should disable breadcrumbs altogether.
- Navigate to a page where you expect to see breadcrumbs based on the theme config value.
  - If a sidebar item has a page its corresponding breadcrumb item should be an anchor that takes you to that page. If it does not have a page its breadcrumb item should just be a `<span>` element, not clickable.
- Change the viewport to a mobile window size, the breadcrumb item font size should be slightly smaller.
- Change the theme between light/dark, the breadcrumb items should change accordingly.

Here is a video demonstrating breadcrumbs in action. It includes the `nested` config, and goes over varying levels of nesting, across different versions, in mobile and desktop, and in light/dark mode.

https://user-images.githubusercontent.com/6392049/151714977-7a04fc52-f643-47c2-b1aa-a6e8db5d7c94.mov

Open to feedback on how I could write better tests for this.

## Related PRs

Closes #6096

---

😄 I took some liberties with this PR, and while I hope you like it I fully anticipate changes to be requested. Excited for your feedback!
